### PR TITLE
Add GPUs in CAPX

### PR DESCRIPTION
**Pull request by yannickstruyf3**
**What this PR does / why we need it**:
Add GPU capabilities to CAPX. Includes e2e tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
N/A

**How Has This Been Tested?**:
- Manual testing on GPU cluster.
- e2e runs for GPU
- e2e runs on non-GPU clusters:
```
Ran 14 of 32 Specs in 1844.147 seconds
SUCCESS! -- 14 Passed | 0 Failed | 0 Pending | 18 Skipped
```

```
LABEL_FILTERS="gpu" RUN_VALIDATION_TESTS_ONLY=true make test-e2e-calico
Ran 6 of 32 Specs in 857.520 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 26 Skipped
PASS
```

**Special notes for your reviewer**:
N/A

**Release note**:

```release-note
- Added GPU support for CAPX clusters
```

### CODE REVIEW

**Review by tuxtof — APPROVED** [review]
functional test are OK
LGTM

File: api/v1beta1/nutanix_types.go [hunk 1]
```diff
@@ -71,3 +80,18 @@ type NutanixCategoryIdentifier struct {
...
+
+type NutanixGPU struct {
+	// Type is the identifier type to use for this resource.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=deviceId;name
+	Type NutanixGPUIdentifierType `json:"type"`
+
+	// deviceID is the id of the GPU entity.
+	// +optional
+	DeviceId *int64 `json:"deviceId,omitempty"`
```
Comment (thunderboltsid) [inline api/v1beta1/nutanix_types.go:92]: nit: can the json property be `device_id` instead?

File: api/v1beta1/nutanix_types.go [hunk 2]
```diff
@@ -35,6 +38,12 @@ const (
 	// NutanixBootTypeUEFI is a resource identifier identifying the UEFI boot type for virtual machines.
 	NutanixBootTypeUEFI NutanixBootType = "uefi"
 
+	// NutanixGPUIdentifierName is a resource identifier identifying a GPU by Name.
+	NutanixGPUIdentifierName NutanixGPUIdentifierType = "name"
+
+	// NutanixGPUIdentifierDeviceID is a resource identifier identifying a GPU using device ID.
+	NutanixGPUIdentifierDeviceID NutanixGPUIdentifierType = "deviceId"
```
Comment (thunderboltsid) [inline api/v1beta1/nutanix_types.go:45]: nit: consider using deviceID or device_id; deviceId seems funky (although not for any particular reason, hence a nit).

File: api/v1beta1/nutanixmachine_types.go [hunk 3]
```diff
@@ -89,6 +89,10 @@ type NutanixMachineSpec struct {
 	// that holds configuration details.
 	// +optional
 	BootstrapRef *corev1.ObjectReference `json:"bootstrapRef,omitempty"`
+
+	// List of GPU devices that need to be added to the machines.
```
Comment (thunderboltsid) [inline api/v1beta1/nutanixmachine_types.go:93]: Add 
```
// +optional
```

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixmachinetemplates.yaml [hunk 4]
```diff
@@ -403,6 +403,29 @@ spec:
                         required:
                         - type
                         type: object
+                      gpus:
+                        description: List of GPU devices that need to be added to
+                          the machines.
+                        items:
+                          properties:
+                            deviceId:
```
Comment (deepakm-ntnx) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixmachinetemplates.yaml:411]: can we add a sample yaml in tests with data populated for deviceId. also how does a user gets this id to pass? nvidia smi? or some acli/ncli or ui?

File: controllers/helpers.go [hunk 5]
```diff
@@ -642,3 +644,77 @@ func hasPEClusterServiceEnabled(peCluster *nutanixClientV3.ClusterIntentResponse
...
+// GetGPUDeviceIdList returns a list of GPU device IDs for the given list of GPUs
+func GetGPUList(ctx context.Context, client *nutanixClientV3.Client, gpus []infrav1.NutanixGPU, peUUID string) ([]*nutanixClientV3.VMGpu, error) {
+	resultGPUs := make([]*nutanixClientV3.VMGpu, 0)
+	for _, gpu := range gpus {
+		foundGPU, err := GetGPU(
+			ctx,
+			client,
+			peUUID,
+			gpu,
+		)
```
Comment (thunderboltsid) [inline controllers/helpers.go:657]: nit: this can be compressed into one line.

File: api/v1beta1/nutanix_types.go (hunk 1 above)
Comment (yannickstruyf3) [reply api/v1beta1/nutanix_types.go:92]: We don't use _ in any other property. Will rename to deviceID instead.

File: api/v1beta1/nutanix_types.go (hunk 2 above)
Comment (yannickstruyf3) [reply api/v1beta1/nutanix_types.go:45]: renaming to deviceID

File: controllers/helpers.go (hunk 5 above)
Comment (yannickstruyf3) [reply controllers/helpers.go:657]: Fixed

File: api/v1beta1/nutanixmachine_types.go (hunk 3 above)
Comment (yannickstruyf3) [reply api/v1beta1/nutanixmachine_types.go:93]: Field already has +kubebuilder:validation:Optional

File: test/e2e/test_helpers.go [hunk 6]
```diff
@@ -232,6 +236,60 @@ func (t testHelper) createUUIDProjectNMT(ctx context.Context, clusterName, names
...
+
+func (t testHelper) findGPU(ctx context.Context, gpuName string) *prismGoClientV3.GPU {
+	clusterVarValue := t.getVariableFromE2eConfig(clusterVarKey)
+
+	clusterUUID, err := controllers.GetPEUUID(ctx, t.nutanixClient, &clusterVarValue, nil)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(clusterUUID).ToNot(BeNil())
+	allGpus, err := controllers.GetGPUsForPE(ctx, t.nutanixClient, clusterUUID)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(allGpus).ToNot(HaveLen(0))
```
Comment (thunderboltsid) [inline test/e2e/test_helpers.go:265]: IMO, this assertion should fail with a message stating this test relies on GPU being associated with PE and is expected to fail on PE without GPU.

**Review by tuxtof — APPROVED** [review]
LGTM

**Review by thunderboltsid — APPROVED** [review]

**Pull request merged**